drivers/usbmisc/fusb302.c

```diff
@@ -0,0 +1,1941 @@
+      break;
+    case TOGGS_SRC_CC2:
+      {
+        enableccpd(priv, CC1_PULLDOWN, false);
+        enableccpd(priv, CC2_PULLDOWN, false);
+        enableccpu(priv, CC1_PULLUP, false);
+        enableccpu(priv, CC2_PULLUP, true);
+        enablevconn(priv, CC1_VCONN, true);
+        enablevconn(priv, CC2_VCONN, false);
+        enableccmeas(priv, CC1_MEASURE, false);
+        enableccmeas(priv, CC2_MEASURE, true);
+        setmdac(priv, SRC_CURRENT_DEFAULT);
+      }
+      break;
+    case TOGGS_SNK_CC1:
+      {
+        enablevconn(priv, CC1_VCONN, false);
+        enablevconn(priv, CC2_VCONN, false);
+        enableccpu(priv, CC1_PULLUP, false);
+        enableccpd(priv, CC2_PULLUP, false);
+        enablevconn(priv, CC1_VCONN, false);
+        enablevconn(priv, CC2_VCONN, false);
+        enableccpd(priv, CC1_PULLDOWN, true);
+        enableccpd(priv, CC2_PULLDOWN, false);
+      }
+      break;
+    case TOGGS_SNK_CC2:
+      {
+        enablevconn(priv, CC1_VCONN, false);
+        enablevconn(priv, CC2_VCONN, false);
+        enableccpu(priv, CC1_PULLUP, false);
+        enableccpu(priv, CC2_PULLUP, false);
+        enablevconn(priv, CC1_VCONN, false);
+        enablevconn(priv, CC2_VCONN, false);
+        enableccpd(priv, CC1_PULLDOWN, false);
+        enableccpd(priv, CC2_PULLDOWN, true);
+      }
+      break;
+    case TOGGS_AUDIO_ACCESSORY:
+        enableccpd(priv, CC1_PULLDOWN, false);
+        enableccpd(priv, CC2_PULLDOWN, false);
+        enableccpu(priv, CC1_PULLUP, true);
+        enableccpu(priv, CC2_PULLUP, false);
+        enablevconn(priv, CC1_VCONN, false);
+        enablevconn(priv, CC2_VCONN, true);
+        enableccmeas(priv, CC1_MEASURE, true);
+        enableccmeas(priv, CC2_MEASURE, false);
+        setmdac(priv, SRC_CURRENT_DEFAULT);
+      break;
+    default:
+      fusb302_err("ERROR: invalid TOGGS VAL read:%d\n", toggsval);
+      break;
+  }
+}
+
+/****************************************************************************
+ * Name: fusb302_worker
+ *
+ * Description:
+ *   Worker task to deal with device state machine
+ *
+ * Input Parameters:
+ *   arg - Pointer to device
+ *
+ * Returned Value:
+ *   OK
+ *
+ ****************************************************************************/
+
+static void fusb302_worker(FAR void *arg)
+{
+  FAR struct fusb302_dev_s    *priv = (FAR struct fusb302_dev_s *)arg;
+  FAR struct fusb302_config_s *config;
+  int                         ret;
+  int                         regval;
+  int                         toggsval;
+  int                         interrupt;
+  int                         interrupta;
+
+  DEBUGASSERT(priv != NULL);
+
+  config = priv->config;
+  DEBUGASSERT(config != NULL);
+
+  /* get exclusive access to the driver data structure */
+
+  do
+    {
+      ret = nxmutex_lock(&priv->devlock);
+
+      /* would only fail if something cancelled the worker thread? */
+
+      DEBUGASSERT(ret == OK || ret == -ECANCELED);
+    }
+  while (ret < 0);
```

Comment:
Why to add this do/while? Look at include/nuttx/mutex.h there is already a for loop to take the mutex

Reply:
This was based on the pre-existing FUSB301 and 302 drivers but a year ago before everything moved to using mutex rather than semaphores. It's left over from that and was unspotted by my when I updated my driver - and is unnecessary as you say. Thanks!